kore/src/Kore/Strategies/Goal.hs

```diff
@@ -40,94 +101,136 @@ See also: 'Strategy.pickFinal', 'extractUnproven'
+proven
+    :: forall goal a
+    .  Goal goal
+    => ProofState.ProofState a ~ ProofState goal a
+    => Strategy.ExecutionGraph (ProofState goal a) (Rule goal)
+    -> Bool
 proven = Foldable.null . unprovenNodes
 
-{- | The primitive transitions of the all-path reachability proof strategy.
- -}
-data Prim rule
-    = CheckProven
-    -- ^ End execution on this branch if the state is 'Proven'.
-    | CheckGoalRemainder
-    -- ^ End execution on this branch if the state is 'GoalRemainder'.
-    | ResetGoal
-    -- ^ Mark all goals rewritten previously as new goals.
-    | Simplify
-    | RemoveDestination
-    | TriviallyValid
-    | DerivePar [rule]
-    | DeriveSeq [rule]
-    deriving (Show)
-
 class Goal goal where
     data Rule goal
+    type Prim goal
+    type ProofState goal a
 
-    -- | Remove the destination of the goal.
-    removeDestination
-        :: MonadSimplify m
-        => goal
-        -> Strategy.TransitionT (Rule goal) m goal
-
-    simplify
+    transitionRule
         :: MonadSimplify m
-        => goal
-        -> Strategy.TransitionT (Rule goal) m goal
+        => Prim goal
+        -> ProofState goal goal
+        -> Strategy.TransitionT (Rule goal) m (ProofState goal goal)
 
-    isTriviallyValid :: goal -> Bool
+    strategy
+        :: [goal]
+        -> [Rule goal]
+        -> [Strategy (Prim goal)]
 
-    isTrusted :: goal -> Bool
+instance (SimplifierVariable variable) => Goal (OnePathRule variable) where
 
-    -- | Apply 'Rule's to the goal in parallel.
-    derivePar
-        :: MonadSimplify m
-        => [Rule goal]
-        -> goal
-        -> Strategy.TransitionT (Rule goal) m (ProofState goal)
+    newtype Rule (OnePathRule variable) =
+        OnePathRewriteRule { unRule :: RewriteRule variable }
+        deriving (GHC.Generic, Show, Unparse)
 
-    -- | Apply 'Rule's to the goal in sequence.
-    deriveSeq
-        :: MonadSimplify m
-        => [Rule goal]
-        -> goal
-        -> Strategy.TransitionT (Rule goal) m (ProofState goal)
-
-transitionRule
-    :: (MonadSimplify m, Goal goal)
-    => Prim (Rule goal)
-    -> ProofState goal
-    -> Strategy.TransitionT (Rule goal) m (ProofState goal)
-transitionRule = transitionRuleWorker
+    type Prim (OnePathRule variable) =
+        ProofState.Prim (Rule (OnePathRule variable))
+
+    type ProofState (OnePathRule variable) a =
+        ProofState.ProofState a
+
+    transitionRule =
+        transitionRule0
+            simplify
+            removeDestination
+            isTriviallyValid
+            derivePar
+            deriveSeq
+
+    strategy goals rules =
+        onePathFirstStep rewrites
+        : repeat
+            ( onePathFollowupStep
+                coinductiveRewrites
+                rewrites
+            )
+      where
+        rewrites = rules
+        coinductiveRewrites =
+            OnePathRewriteRule
+            . RewriteRule
+            . getOnePathRule
+            <$> goals
+
+transitionRule0
```

Comment:
The name of this function did not clarify its purpose for me. Could we rename this to `transitionRuleTemplate`? I would also prefer if it took a record of functions as its argument; this would be easier to use and clearer to document.

Because this depends on `ProofState.ProofState`, should it go live with that type?

Reply:
> Because this depends on `ProofState.ProofState`, should it go live with that type?

Sorry, I'm not sure I understand what you mean here. Would you prefer I remove the type equality constraint in that function and use `ProofState.ProofState` directly?

Reply:
I just meant to move it into the same module as `ProofState`, but it's not important.